File: Frontend/Apple/Cookey/Services/TrustedKeyStore.swift
import Foundation

// MARK: - Verification State

enum KeyVerificationState: Equatable {
    case trusted(fingerprint: String)
    case keyChanged(oldFingerprint: String, newFingerprint: String)
    case knownKeyNewDevice(fingerprint: String)
    case firstTime(fingerprint: String)
}

// MARK: - Model

struct TrustedKey: Codable, Equatable {
    let deviceID: String
    let publicKeyBase64: String
    let fingerprint: String
    let firstTrustedAt: Date
    var lastSeenAt: Date
    var label: String?
}

// MARK: - Store

enum TrustedKeyStore {
    // MARK: Public

    static func verify(deviceID: String, publicKeyBase64: String) -> KeyVerificationState {
        let fingerprint = (try? KeyFingerprint.compute(fromX25519PublicKeyBase64: publicKeyBase64)) ?? publicKeyBase64
        let entries = load()

        if let existing = entries.first(where: { $0.deviceID == deviceID }) {
            if existing.publicKeyBase64 == publicKeyBase64 {
                touchLastSeen(deviceID: deviceID)
                return .trusted(fingerprint: fingerprint)
            }
            return .keyChanged(oldFingerprint: existing.fingerprint, newFingerprint: fingerprint)
        }

        if entries.contains(where: { $0.publicKeyBase64 == publicKeyBase64 }) {
            return .knownKeyNewDevice(fingerprint: fingerprint)
        }

        return .firstTime(fingerprint: fingerprint)
    }

    static func trust(deviceID: String, publicKeyBase64: String, fingerprint: String) {
        var entries = load()
        let now = Date()

        if let index = entries.firstIndex(where: { $0.deviceID == deviceID }) {
            var entry = entries[index]
            entry = TrustedKey(
                deviceID: deviceID,
                publicKeyBase64: publicKeyBase64,
                fingerprint: fingerprint,
                firstTrustedAt: entry.firstTrustedAt,
                lastSeenAt: now,
                label: entry.label
            )
            entries[index] = entry
        } else {
            entries.append(TrustedKey(
                deviceID: deviceID,
                publicKeyBase64: publicKeyBase64,
                fingerprint: fingerprint,
                firstTrustedAt: now,
                lastSeenAt: now,
                label: nil
            ))
        }

        save(entries)
    }

    static func remove(deviceID: String) {
        var entries = load()
        entries.removeAll { $0.deviceID == deviceID }
        save(entries)
    }

    static func allTrusted() -> [TrustedKey] {
        load()
    }

    static func moveItem(from sourceIndex: Int, to destinationIndex: Int) {
        var entries = load()
        guard sourceIndex >= 0, sourceIndex < entries.count,
              destinationIndex >= 0, destinationIndex < entries.count
        else { return }
        let item = entries.remove(at: sourceIndex)
        entries.insert(item, at: destinationIndex)
        save(entries)
    }

    private static func touchLastSeen(deviceID: String) {
        var entries = load()
        guard let index = entries.firstIndex(where: { $0.deviceID == deviceID }) else { return }
        entries[index].lastSeenAt = Date()
        save(entries)
    }

    // MARK: Private

    private static let lock = NSLock()

    private static var storageURL: URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let directory = appSupport.appendingPathComponent("wiki.qaq.cookey.app", isDirectory: true)
        return directory.appendingPathComponent("trusted_clis.json")
    }

    private static func load() -> [TrustedKey] {
        lock.lock()
        defer { lock.unlock() }

        guard let data = try? Data(contentsOf: storageURL) else {
            return []
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode([TrustedKey].self, from: data)) ?? []
    }

    private static func save(_ entries: [TrustedKey]) {
        lock.lock()
        defer { lock.unlock() }

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]

        guard let data = try? encoder.encode(entries) else { return }

        let fileManager = FileManager.default
        let directory = storageURL.deletingLastPathComponent()

        if !fileManager.fileExists(atPath: directory.path) {
            try? fileManager.createDirectory(at: directory, withIntermediateDirectories: true, attributes: [
                .posixPermissions: 0o700,
            ])
        }

        let tempURL = directory.appendingPathComponent(UUID().uuidString + ".tmp")

        do {
            try data.write(to: tempURL, options: [.atomic])
            try fileManager.setAttributes([.posixPermissions: 0o600], ofItemAtPath: tempURL.path)
            if fileManager.fileExists(atPath: storageURL.path) {
                _ = try fileManager.replaceItemAt(storageURL, withItemAt: tempURL)
            } else {
                try fileManager.moveItem(at: tempURL, to: storageURL)
            }
        } catch {
            try? fileManager.removeItem(at: tempURL)
        }
    }
}

File: Frontend/Android/app/src/main/java/wiki/qaq/cookey/service/TrustedKeyStore.kt
package wiki.qaq.cookey.service

import android.content.Context
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import wiki.qaq.cookey.crypto.KeyFingerprint
import java.io.File

@Serializable
data class TrustedKey(
    @SerialName("device_id")
    val deviceID: String,
    @SerialName("public_key_base64")
    val publicKeyBase64: String,
    val fingerprint: String,
    @SerialName("first_trusted_at")
    val firstTrustedAt: String,
    @SerialName("last_seen_at")
    var lastSeenAt: String,
    var label: String? = null
)

enum class KeyVerificationState {
    TRUSTED,
    KEY_CHANGED,
    KNOWN_KEY_NEW_DEVICE,
    FIRST_TIME
}

data class KeyVerificationResult(
    val state: KeyVerificationState,
    val fingerprint: String,
    val oldFingerprint: String? = null
)

object TrustedKeyStore {

    private val json = Json { prettyPrint = true; ignoreUnknownKeys = true }
    private val lock = Any()

    private fun storeFile(context: Context): File {
        val dir = File(context.filesDir, "wiki.qaq.cookey.app")
        if (!dir.exists()) dir.mkdirs()
        val file = File(dir, "trusted_clis.json")
        file.setReadable(true, true)
        file.setWritable(true, true)
        return file
    }

    private fun loadKeys(context: Context): MutableList<TrustedKey> {
        val file = storeFile(context)
        if (!file.exists()) return mutableListOf()
        return try {
            json.decodeFromString<MutableList<TrustedKey>>(file.readText())
        } catch (_: Exception) {
            mutableListOf()
        }
    }

    private fun saveKeys(context: Context, keys: List<TrustedKey>) {
        val file = storeFile(context)
        file.writeText(json.encodeToString(keys))
    }

    fun verify(
        context: Context,
        deviceID: String,
        publicKeyBase64: String
    ): KeyVerificationResult {
        val fingerprint = KeyFingerprint.compute(publicKeyBase64)
        synchronized(lock) {
            val keys = loadKeys(context)

            // Exact match: same device, same key
            val exactMatch = keys.find { it.deviceID == deviceID && it.publicKeyBase64 == publicKeyBase64 }
            if (exactMatch != null) {
                return KeyVerificationResult(KeyVerificationState.TRUSTED, fingerprint)
            }

            // Same device, different key
            val sameDevice = keys.find { it.deviceID == deviceID }
            if (sameDevice != null) {
                return KeyVerificationResult(
                    KeyVerificationState.KEY_CHANGED,
                    fingerprint,
                    oldFingerprint = sameDevice.fingerprint
                )
            }

            // Same key, different device
            val sameKey = keys.find { it.publicKeyBase64 == publicKeyBase64 }
            if (sameKey != null) {
                return KeyVerificationResult(KeyVerificationState.KNOWN_KEY_NEW_DEVICE, fingerprint)
            }

            // Brand new
            return KeyVerificationResult(KeyVerificationState.FIRST_TIME, fingerprint)
        }
    }

    fun trust(context: Context, deviceID: String, publicKeyBase64: String) {
        val fingerprint = KeyFingerprint.compute(publicKeyBase64)
        val now = java.time.Instant.now().toString()
        synchronized(lock) {
            val keys = loadKeys(context)

            // Remove old entry for same device if exists
            keys.removeAll { it.deviceID == deviceID }

            keys.add(
                TrustedKey(
                    deviceID = deviceID,
                    publicKeyBase64 = publicKeyBase64,
                    fingerprint = fingerprint,
                    firstTrustedAt = now,
                    lastSeenAt = now
                )
            )
            saveKeys(context, keys)
        }
    }

    fun updateLastSeen(context: Context, deviceID: String) {
        val now = java.time.Instant.now().toString()
        synchronized(lock) {
            val keys = loadKeys(context)
            keys.find { it.deviceID == deviceID }?.lastSeenAt = now
            saveKeys(context, keys)
        }
    }

    fun allKeys(context: Context): List<TrustedKey> {
        synchronized(lock) {
            return loadKeys(context)
        }
    }

    fun removeKey(context: Context, deviceID: String) {
        synchronized(lock) {
            val keys = loadKeys(context)
            keys.removeAll { it.deviceID == deviceID }
            saveKeys(context, keys)
        }
    }
}
